### PR TITLE
refactor: replace sanitizer with dompurify

### PR DESCRIPTION
**Pull request by Mnau-mnau**
Sanitizer is no longer supported library, so it is replaced by dompurify
[gist test & lib comparison](https://gist.github.com/Mnau-mnau/ad5cb202bdc2716ab3bfce8e3a5db110)

BREAKING CHANGE:

 * make options.sanitize default OFF
 * make options.sanitize a function instead of a boolean
 * use jsdom in tests (for dompurify to work)

Bundlers should be able to tree-shake the `dompurify` out when the
exported sanitize function is never used by the app.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,11 +21,12 @@
   "author": "Apiary Czech Republic, s.r.o. <support@apiary.io>",
   "license": "MIT",
   "dependencies": {
+    "dompurify": "^1.0.10",
+    "jsdom": "^14.0.0",
```
Comment (opichals) [inline package.json]: `jsdom` should only be a devDependency used in tests. In browsers this should use the browser DOM.

Comment (Mnau-mnau) [reply package.json]: changed

File: lib/sanitizerPolicies.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+function addTabindexToIdPolicy(node) {
```
Comment (opichals) [inline lib/sanitizerPolicies.js]: Please sync with phoenix-ui version of this file.

File: lib/commonmark.js [hunk 3]
```diff
@@ -1,8 +1,9 @@
 'use strict';
 
 var MarkdownIt = require('markdown-it');
-var sanitizer = require('sanitizer');
-var uriPolicy = require('./uriPolicy').uriPolicy;
+var createDOMPurify = require('dompurify');
+var { JSDOM } = require('jsdom');
```
Comment (opichals) [inline lib/commonmark.js]: The use of `jsdom` must be only done for tests. It must not be present in the implementation.

Comment (Mnau-mnau) [reply lib/commonmark.js]: ```jsdom``` moved to ```renderHtml``` for tests

File: lib/sanitizerPolicies.js (hunk 2 above)
Comment (Mnau-mnau) [reply lib/sanitizerPolicies.js]: changed

File: lib/commonmark.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
 'use strict';
 
 var MarkdownIt = require('markdown-it');
-var sanitizer = require('sanitizer');
-var uriPolicy = require('./uriPolicy').uriPolicy;
+var DOMPurify = require('dompurify');
```
Comment (opichals) [inline lib/commonmark.js]:
```suggestion
var dompurify = require('dompurify');
```

File: lib/commonmark.js [hunk 5]
```diff
@@ -1,8 +1,8 @@
...
+var { linkPolicy, addTabindexToIdPolicy } = require('./policies.js');
```
Comment (opichals) [inline lib/commonmark.js]: Let's hide the `dompurify` into a local `./sanitize.js` file completely with no separate policies file.

```suggestion
var sanitize = require('./sanitize.js');
```

Comment (Mnau-mnau) [reply lib/commonmark.js]: moved to sanitize.js

File: lib/commonmark.js (hunk 4 above)
Comment (Mnau-mnau) [reply lib/commonmark.js]: renamed

File: test/description-render-test.js [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 var assert = require('chai').assert;
 
-var markdownRenderer = require('../lib/index.js');
+var { markdownRenderer } = require('./markdownRendererWithDom.js');
```
Comment (opichals) [inline test/description-render-test.js]: The test should deal with the `../lib/index.js` here.  I'd just add global.window as follows:

```suggestion
global.window = (new JSDOM('')).window;
var { markdownRenderer } = require('../lib/index.js');
```

File: lib/commonmark.js [hunk 7]
```diff
@@ -67,7 +66,9 @@ function renderHtml(markdownString, options, cb) {
   }
 
   if (options.sanitize) {
-    result = sanitizer.sanitize(result, uriPolicy);
+    dompurify.addHook('afterSanitizeAttributes', sanitize);
```
Comment (opichals) [inline lib/commonmark.js]: Please move all the `dompurify` use to the `sanitize.js`. Perhaps even nicer would be to expect the `options.sanitize` to be a functor which could be passed from outside. (would export the sanitize function from the `lib/index.js` to handle it. That way the code could be tree-shaken away when not used by the app.

```suggestion
if (typeof options.sanitize === 'function') {
   result = options.sanitize(result);
}
```

File: lib/commonmark.js [hunk 8]
```diff
@@ -25,7 +23,6 @@ var defaultOptions = {
 
 var defaultRenderer = rendererFactory();
 
-
```
Comment (opichals) [inline lib/commonmark.js]: keep the line please

File: lib/sanitize.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+  if (!absoluteUriRegexp.test(href)) { return; }
+
+  // default caja link href sanitize
+  if (!allowedUriSchemesRegexp.test(href)) {
+    node.removeAttribute('href');
+    return;
+  }
+
+  // node.setAttribute('target', '_blank');
```
Comment (opichals) [inline lib/sanitize.js]: drop these comments

File: lib/sanitize.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  linkPolicy,
+  addTabindexToIdPolicy,
+]);
+
+dompurify.addHook('afterSanitizeAttributes', sanitizeHtmlPolicy);
+
+function sanitize(source, config) {
+  return dompurify.sanitize(source, config);
+}
+  // const newResult = DOMPurify.sanitize(result);
```
Comment (opichals) [inline lib/sanitize.js]: drop the commented code

File: lib/sanitize.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
+var dompurify = require('dompurify');
+
+
+function addTabindexToIdPolicy(node) {
```
Comment (opichals) [inline lib/sanitize.js]: We should drop adding `tabIndex` as the `sanitizer` does not do that.

Comment (Mnau-mnau) [reply lib/sanitize.js]: removed

File: lib/sanitize.js (hunk 10 above)
Comment (Mnau-mnau) [reply lib/sanitize.js]: removed

File: lib/sanitize.js (hunk 9 above)
Comment (Mnau-mnau) [reply lib/sanitize.js]: removed

File: lib/commonmark.js (hunk 8 above)
Comment (Mnau-mnau) [reply lib/commonmark.js]: added

File: lib/sanitize.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  if (node.hasAttribute('id')) {
+    node.setAttribute('tabindex', '-1');
+  }
+};
+
+const allowedUriSchemesRegexp = /^(https?|mailto|#)/i;
+const absoluteUriRegexp = /^[a-zA-Z]+:\//;
+
+function linkPolicy(node) {
```
Comment (opichals) [inline lib/sanitize.js]: We need to add tests for this hook.

Comment (Mnau-mnau) [reply lib/sanitize.js]: tests added

File: lib/sanitize.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
+var dompurify = require('dompurify');
+
+
+const allowedUriSchemesRegexp = /^(https?|mailto|#)/i;
```
Comment (patricksmms) [inline lib/sanitize.js]: Just to keep the codebase consistent, going with ES5 all around:
```suggestion
var allowedUriSchemesRegexp = /^(https?|mailto|#)/i;
```

File: lib/sanitize.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+const absoluteUriRegexp = /^[a-zA-Z]+:\//;
```
Comment (patricksmms) [inline lib/sanitize.js]:
```suggestion
var absoluteUriRegexp = /^[a-zA-Z]+:\//;
```

File: lib/sanitize.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+function linkHrefHook(node) {
+  if (node.nodeName !== 'A') { return; }
+
+  const href = node.getAttribute('href') || '';
```
Comment (patricksmms) [inline lib/sanitize.js]:
```suggestion
  var href = node.getAttribute('href') || '';
```

File: test/description-render-test.js [hunk 16]
```diff
@@ -266,11 +269,71 @@ describe('Markdown', function() {
         var expectedHtml, markdownString;
         markdownString = '<img src="/image.jpg" onerror="alert(\'XSS\')" />';
         expectedHtml = '<img src="/image.jpg">\n';
-        markdownRenderer.renderHtml(markdownString, function(error, html) {
+        markdownRenderer.renderHtml(markdownString, { sanitize: sanitize }, function(error, html) {
           assert.strictEqual(html, expectedHtml);
           done(error);
         });
       });
+      context('given <a> tag', function () {
```
Comment (patricksmms) [inline test/description-render-test.js]: Please add a new line before this:
```suggestion

      context('given <a> tag', function () {
```

File: lib/sanitize.js (hunk 13 above)
Comment (Mnau-mnau) [reply lib/sanitize.js]: changed

File: lib/sanitize.js (hunk 15 above)
Comment (Mnau-mnau) [reply lib/sanitize.js]: changed

File: test/description-render-test.js (hunk 16 above)
Comment (Mnau-mnau) [reply test/description-render-test.js]: changed

File: lib/sanitize.js (hunk 14 above)
Comment (Mnau-mnau) [reply lib/sanitize.js]: changed

**Pull request merged**